[EGL] Fail early if EGL version is less than 1.5

Steelskin opened the pull request:
* Add an explicit error message indicating why the `wxGLCanvas` failed to initialize on systems where the EGL runtime version is older than 1.5.
* Change `wxGLCanvasEGL::GetDisplay()` to use `eglGetDisplay()` rather than `eglGetPlatformDisplay()` when the latter is not available. This is necessary because this has to be called before EGL is initialized, which is the point where we can know the EGL runtime version. `eglGetPlatformDisplay()` is only available on EGL 1.5 and greater.

Bug: #22325

TEST=Tested the OpenGL samples on Wayland with EGL 1.5. Hard-coded the EGL version number to an older version to ensure the samples were failing with the expected error message.

Steelskin committed:
[EGL] Fail early if EGL version is less than 1.5

* Add an explicit error message indicating why the `wxGLCanvas` failed
  to initialize on systems where the EGL runtime version is older than
  1.5.
* Change `wxGLCanvasEGL::GetDisplay()` to use `eglGetDisplay()` rather
  than `eglGetPlatformDisplay()`. This is necessary because this has to
  be called before EGL is initialized, which is the point where we can
  know the EGL runtime version. `eglGetPlatformDisplay()` is only
  available on EGL 1.5 and greater.

Bug: #22325

TEST=Tested the OpenGL samples on Wayland with EGL 1.5. Hard-coded the
EGL version number to an older version to ensure the samples were
failing with the expected error message.

swt2c commented:
Thanks for the PR!  After researching this a little bit, it seems that using `eglGetPlatformDisplay()` is preferred over using `eglGetDisplay()`.  It seems that it is possible to determine whether `eglGetPlatformDisplay()` is available by checking the EGL extensions?  See for example:

https://github.com/WebKit/WebKit/blob/5e7400cc3d50fe4b4657242113ef03de1868bec3/Source/WebCore/platform/graphics/x11/PlatformDisplayX11.cpp#L137

vadz commented on src/unix/glegl.cpp:
```diff
@@ -628,11 +615,26 @@ EGLConfig *wxGLCanvasEGL::InitConfig(const wxGLAttributes& dispAttrs)
+    {
+        // Ignore the return value here, we cannot recover at this point.
+        eglTerminate(dpy);
+        wxFAIL_MSG(wxString::Format(
+            "EGL version is %d.%d. EGL version 1.5 or greater is required.",
+            egl_major, egl_minor));
```
I wonder if this shouldn't be `wxLogError()`, so that it is reported to the end user in production builds (where `wxFAIL_MSG()` does nothing by default), as it might be something they can fix or at least tell the developer about.

Steelskin replied:
I changed it. I think it's going to be more useful down the line, when we can fallback more graciously for older versions. Anyway it doesn't change much right now since we end up calling `wxFAIL_MSG()` down the line.

vadz reviewed (commented):
Thanks! I'm not merging this yet because I'm not sure about `eglGetDisplay()` vs `eglGetPlatformDisplay()` (when in doubt I'd rather not change the existing code), but otherwise this looks good to me.

Steelskin committed:
[EGL] Use eglGetPlatformDisplay() if available

* Restored the code using `eglGetPlatformDisplay()` rather than
  `eglGetDisplay()`, if the former is available. This is done by
  checking for available EGL extensions.
* Fixed a couple of code style issues.
* Fail with `wxLogError()` rather than `wxFAIL_MSG()` when the runtime
  EGL version is too old. Note that `wxFAIL_MSG()` still ends up being
  called down the line due to initialization failure.

Steelskin commented:
I changed the code to check for the presence of the extension, rather than always using `eglGetDisplay()`.

swt2c commented on src/unix/glegl.cpp:
```diff
@@ -367,6 +367,17 @@ bool wxGLCanvasEGL::InitVisual(const wxGLAttributes& dispAttrs)
+    static const bool s_EGLHasPlatformDisplay =
+        wxGLCanvasBase::IsExtensionInList(
+            eglQueryString(nullptr, EGL_EXTENSIONS),
+            "EGL_EXT_platform_base");
```
I think technically the extension for `eglGetPlatformDisplay()` should be `EGL_KHR_platform_base`.  `EGL_EXT_platform_base` seems to be for `eglGetPlatformDisplayEXT()`.

Steelskin replied:
I am testing on Ubuntu 22.04 running under WSL, and this is what I am getting from `eglQueryString()`:

```
EGL_EXT_device_base
EGL_EXT_device_enumeration
EGL_EXT_device_query
EGL_EXT_platform_base
EGL_KHR_client_get_all_proc_addresses
EGL_EXT_client_extensions
EGL_KHR_debug
EGL_EXT_platform_device
EGL_EXT_platform_wayland
EGL_KHR_platform_wayland
EGL_EXT_platform_x11
EGL_KHR_platform_x11
EGL_EXT_platform_xcb
EGL_MESA_platform_gbm
EGL_KHR_platform_gbm
EGL_MESA_platform_surfaceless
```

In fact, I cannot find `EGL_KHR_platform_base` anywhere in the EGL spec. But I think this code is wrong either way, as `EGL_EXT_platform_base` is defined for EGL 1.4+, and `eglGetPlatformDisplay()` is EGL 1.5. I'm going to spend some more time on this.

Steelskin replied:
Okay, I ended up using `eglGetProcAddress()` instead and added `eglGetPlatformDisplayEXT()` as an intermediate fallback. I think this should work either way but I don't have any 1.4 hardware to test.

Steelskin committed:
[EGL] Use eglGetProcAddress() to get EGLDisplay

This modifies the EGLDisplay initialization to pick the appropriate
function at runtime, depending on what is available. The preferred
function is `eglGetPlatformDisplay()`, which is only available on EGL
1.5+. As a fallback, `eglGetPlatformDisplayEXT()` is used. This is
available on some platforms using EGL 1.4, but is no longer available on
EGL 1.5. If all else fails, use the older `eglGetDisplay()` API.

swt2c commented on src/unix/glegl.cpp:
```diff
@@ -364,9 +364,51 @@ bool wxGLCanvasEGL::InitVisual(const wxGLAttributes& dispAttrs)
+
+// Returns the appropriate function pointer for "eglGetPlatformDisplay()" or
+// "eglGetPlatformDisplayEXT()" or nullptr if neither is available at runtime.
+static GetPlatformDisplayFunc get_edl_platform_display_func()
```
Presumably you meant to name this `get_egl_platform_display_func` ?

Steelskin replied:
I ended up removing this in favor of embedding the static variable in `GetDisplay()` since I now need to call `wxGLCanvasBase::IsExtensionInList()`, which is protected. I can't use `wxGLCanvasBase::IsExtensionSupported()` from `wxGLCanvasEGL::GetDisplay()` because the former calls the latter.

swt2c commented on src/unix/glegl.cpp:
```diff
@@ -364,9 +364,51 @@ bool wxGLCanvasEGL::InitVisual(const wxGLAttributes& dispAttrs)
+    {
+        const GetPlatformDisplayFunc getPlatformDisplayFunc =
+            reinterpret_cast<GetPlatformDisplayFunc>(
+                eglGetProcAddress("eglGetPlatformDisplay"));
```
It looks like just calling `eglGetProcAddress()` might not be enough.  :-(  From the [documentation](https://registry.khronos.org/EGL/sdk/docs/man/html/eglGetProcAddress.xhtml):

> A non-NULL return value does not guarantee that an extension function is actually supported at runtime. The client must also make a corresponding query, such as glGetString(GL_EXTENSIONS) for OpenGL and OpenGL ES extensions; vgGetString(VG_EXTENSIONS) for OpenVG extensions; eglQueryString(display, EGL_EXTENSIONS); or query the EGL or client API version for non-extension functions, to determine if a function is supported by EGL or a specific client API context.

So it seems it might be necessary to call `eglQueryString` also.

swt2c replied:
Here is how GTK seems to do it, for reference:
https://gitlab.gnome.org/GNOME/gtk/-/blob/main/gdk/gdkdisplay.c#L1488

Steelskin replied:
I really think the `EGL_KHR_platform_base` part is incorrect. I can at least attest of one 1.5 EGL implementation that doesn't use it. It is also nowhere to be found in the EGL documentation.

swt2c replied:
> I really think the `EGL_KHR_platform_base` part is incorrect. I can at least attest of one 1.5 EGL implementation that doesn't use it. It is also nowhere to be found in the EGL documentation.

You might be right about `EGL_KHR_platform_base`.  [This](https://registry.khronos.org/EGL/api/egl.xml) seems to suggest that `EGL_EXT_platform_base` should be used instead.

vadz replied:
So is the current version ok to merge, @swt2c?

swt2c replied:
I still think that checks for `EGL_EXT_platform_base` in the query string should be added too.

Also fix the presumed typo I pointed out.

Steelskin replied:
I am now querying for `EGL_EXT_platform_base` before trying `eglGetProcAddress()`. Please, let me know what you think.

swt2c replied:
This looks OK to me now.

My only nitpick would be the lambda function as this can't be backported as-is, but it's not a dealbreaker.  :-)

Steelskin committed:
[EGL] Query for extension to get display

This adds a query for the `EGL_EXT_platform_base` extension before
attempting to get either `eglGetPlatformDisplay` or
`eglGetPlatformDisplayEXT` function address. This seems to be the
recommended way of getting either function, per the EGL spec. See
https://registry.khronos.org/EGL/api/egl.xml for reference.

Steelskin committed:
[EGL] Move extension query to a separate variable

It is necessary to call `wxGLCanvasBase::IsExtensionInList()` outside of
the lambda to work around a compiler bug in GCC 4.8.

vadz committed:
Slightly simplify s_eglGetPlatformDisplay initialization

No real changes, just avoid using the IIFE which doesn't seem very
useful here.

Also explicitly list all enum elements in a switch so that we get a
compiler warning if a new one is ever added.

vadz commented:
Sorry, but I agree with Scott here: this lambda doesn't seem to be really useful, so I'd rather avoid it as I did in the latest commit.

If you don't mind it, I'll (squash?) merge this and probably backport to 3.2 too -- please let me know if you do. Thanks!

Steelskin commented:
@vadz I do not mind the change. The main difference is that the initialization is no longer thread-safe (C++ guarantees const static initializers are only called once), but I don't think that matters. Thanks for following up!

vadz commented:
> @vadz I do not mind the change. The main difference is that the initialization is no longer thread-safe (C++ guarantees const static initializers are only called once), but I don't think that matters. Thanks for following up!

Everything GUI-related in wxWidgets can only be done from the main GUI thread anyhow, so, indeed, I don't think it matters.

I'll merge it when the CI builds finish, thanks again!
